source/include/actbl2.h

```diff
@@ -2265,6 +2265,215 @@ typedef struct nfit_device_handle
 
 
+/*******************************************************************************
+ *
+ * NHLT - Non HDAudio Link Table
+ *        Version 1
+ *
+ ******************************************************************************/
+
+typedef struct acpi_table_nhlt
+{
+    ACPI_TABLE_HEADER              Header;          /* Common ACPI table header */
+    UINT8                          EndpointsCount;
+    /*
+     * ACPI_NHLT_ENDPOINT          Endpoints[];
+     * ACPI_NHLT_CONFIG            OEDConfig;
+     */
+
+} ACPI_TABLE_NHLT;
+
+typedef struct acpi_nhlt_endpoint
+{
+    UINT32                         Length;
+    UINT8                          LinkType;
+    UINT8                          InstanceId;
+    UINT16                         VendorId;
+    UINT16                         DeviceId;
+    UINT16                         RevisionId;
+    UINT32                         SubsystemId;
+    UINT8                          DeviceType;
+    UINT8                          Direction;
+    UINT8                          VirtualBusId;
+    /*
+     * ACPI_NHLT_CONFIG            DeviceConfig;
+     * ACPI_NHLT_FORMATS_CONFIG    FormatsConfig;
+     * ACPI_NHLT_DEVICES_INFO      DevicesInfo;
+     */
+
+} ACPI_NHLT_ENDPOINT;
+
+/* Values for LinkType field above */
+
+#define ACPI_NHLT_LINKTYPE_HDA               0
+#define ACPI_NHLT_LINKTYPE_DSP               1
+#define ACPI_NHLT_LINKTYPE_PDM               2
+#define ACPI_NHLT_LINKTYPE_SSP               3
+#define ACPI_NHLT_LINKTYPE_SLIMBUS           4
+#define ACPI_NHLT_LINKTYPE_SDW               5
+#define ACPI_NHLT_LINKTYPE_UAOL              6
+
+/* Values for DeviceId field above */
+
+#define ACPI_NHLT_DEVICEID_DMIC              0xAE20
+#define ACPI_NHLT_DEVICEID_BT                0xAE30
+#define ACPI_NHLT_DEVICEID_I2S               0xAE34
+
+/* Values for DeviceType field above */
+
+/* Device types unique to endpoint of LinkType=PDM */
+#define ACPI_NHLT_DEVICETYPE_PDM             0
+#define ACPI_NHLT_DEVICETYPE_PDM_SKL         1
+/* Device types unique to endpoint of LinkType=SSP */
+#define ACPI_NHLT_DEVICETYPE_BT              0
+#define ACPI_NHLT_DEVICETYPE_FM              1
+#define ACPI_NHLT_DEVICETYPE_MODEM           2
+#define ACPI_NHLT_DEVICETYPE_CODEC           4
+
+/* Values for Direction field above */
+
+#define ACPI_NHLT_DIR_RENDER                 0
+#define ACPI_NHLT_DIR_CAPTURE                1
+
+typedef struct acpi_nhlt_config
+{
+    UINT32                         CapabilitiesSize;
+    UINT8                          Capabilities[];
+
+} ACPI_NHLT_CONFIG;
+
+typedef struct acpi_nhlt_gendevice_config
+{
+    UINT8                          VirtualSlot;
+    UINT8                          ConfigType;
+
+} ACPI_NHLT_GENDEVICE_CONFIG;
+
+/* Values for ConfigType field above */
+
+#define ACPI_NHLT_CONFIGTYPE_GENERIC         0
+#define ACPI_NHLT_CONFIGTYPE_MICARRAY        1
+
+typedef struct acpi_nhlt_micdevice_config
+{
+    UINT8                          VirtualSlot;
```

Comment:
So, this is basically an embedded `struct acpi_nhlt_gendevice_config`, correct? I dunno if we can use it inline (and it will need packed attribute I believe) per ACPICA code style, but at least a comment maybe added as you done elsewhere like

    /* ACPI_NHLT_GENDEVICE_CONFIG */
    ... VirtualSlot;
    ...

Reply:
Yes, you're correct.

My vote goes for the comment option. I believe that these structs are so thin that adding a "base" element would complicate accessing their fields unnecessarily.

For example, obtaining `ConfigType` for instance of `struct acpi_nhlt_vendor_micdevice_config` would take the form of: `vendor_mic.Base.Base.ConfigType`. As there is a friendly union already (below), I don't think inlining is a bad take.

Reply:
Hmm.. now when I think about it, reader after seeing the comment, may think there are two sets of fields rather than one. Comments tailing description of several structs above make it clear that given descriptor is followed by other data types. Here it's not the case. Nothing "hidden" is following or prepending the device_configs.

I see that there are other examples of such constructs in `actbl2.h`: `struct acpi_ivrs_hardware_10` and `struct acpi_ivrs_hardware_11`.

Reply:
May be elaborated comment? Oh, well, I leave it to ACPICA maintainers and you. The whole thing looks already good enough to me!